src/protocol/transport_service.rs

```diff
@@ -467,15 +468,7 @@ impl TransportService {
     ///
     /// The list is filtered for duplicates and unsupported transports.
     pub fn add_known_address(&mut self, peer: &PeerId, addresses: impl Iterator<Item = Multiaddr>) {
-        let addresses: HashSet<Multiaddr> = addresses
-            .filter_map(|address| {
-                if !std::matches!(address.iter().last(), Some(Protocol::P2p(_))) {
-                    Some(address.with(Protocol::P2p(Multihash::from_bytes(&peer.to_bytes()).ok()?)))
-                } else {
-                    Some(address)
-                }
-            })
-            .collect();
+        let addresses = ensure_address_with_peer(addresses.into_iter(), *peer);
```

Comment:
There's one tiny issue with this approach: we were previously collecting into a `HashSet`, which deduplicates items. Whereas using the Vec by default, we could have duplicates here